simpervisor/process.py

```diff
@@ -17,6 +19,121 @@ class KilledProcessError(Exception):
     """
 
 
+class Process:
+    """
+    Abstract class to start, wait and send signals to running processes in a OS agnostic way
+    """
+
+    # Protected data members only accessible by derived classes.
+    _proc_cmd = None
+    _proc_kwargs = None
+    _proc = None
+
+    def __init__(self, *cmd, **kwargs):
+        self._proc_cmd = cmd
+        self._proc_kwargs = kwargs
+
+    async def start(self):
+        """
+        Start the process
+        """
+        raise NotImplementedError
+
+    async def wait(self):
+        """
+        Wait for the process to terminate and return the process exit code.
+        """
+        raise NotImplementedError
+
+    def get_kill_signal(self):
+        """
+        Returns the preferred OS signal to kill the process.
+        """
+        raise NotImplementedError
+
+    def send_signal(self, signum):
+        """
+        Send the OS signal to the process.
+        """
+        if self._proc:
+            self._proc.send_signal(signum)
+
+    @property
+    def pid(self):
+        if self._proc:
+            return self._proc.pid
+
+    @property
+    def returncode(self):
+        if self._proc:
+            return self._proc.returncode
+
+
+class POSIXProcess(Process):
+    """
+    A process that uses asyncio-subprocess API to start and wait.
+    """
+
+    async def start(self):
+        """
+        Start the process using asyncio.create_subprocess_exec API
+        """
+        self._proc = await asyncio.create_subprocess_exec(
+            *self._proc_cmd, **self._proc_kwargs
+        )
+
+    async def wait(self):
+        """
+        Wait for the process to stop and return the process exit code.
+        """
+        return await self._proc.wait()
+
+    def get_kill_signal(self):
+        """
+        Returns the OS signal used for kill the child process.
+        """
+        return signal.SIGKILL
+
+
+class WindowsProcess(Process):
+    """
+    A process that uses subprocess API to start and wait (uses busy polling).
+    """
+
+    async def start(self):
+        """
+        Starts the process using subprocess.Popen API
+        """
+        self._proc = subprocess.Popen(list(self._proc_cmd), **self._proc_kwargs)
+
+    async def wait(self):
+        """
+        Wait for the process to stop and return the process exit code.
+
+        subprocess.Popen.wait() is a blocking call which can cause the asyncio
+        event loop to remain blocked until the child process is terminated.
+
+        To circumvent this behavior, we use busy polling with asyncio.sleep to check
+        whether the child process is alive or not and keeping the asyncio event
+        loop running.
+
+        See https://github.com/jupyter/jupyter_client/blob/main/jupyter_client/provisioning/local_provisioner.py#L54_L55 for similar use.
+        """
+        while self._proc.poll() is None:
+            await asyncio.sleep(0.1)
```

Comment:
```suggestion
            await asyncio.sleep(0.5)
```

I think it's ok to wait 500ms between checks rather than 100ms.